### ISSUE TITLE
💡 Yield Overrides

### ISSUE THREAD

**Issue by arcanis** [microsoft/TypeScript#43632]
# Suggestion

## 🔍 Search Terms

yield, any, generator, saga

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

A new `YieldOverride` builtin generic type would instruct TypeScript to ignore the actual generator type in a `yield` expression, and to instead use whatever is used as generic parameter. The following would thus be true:

```ts
declare const val: YieldOverride<boolean>;

function* generator() {
  const res = yield val;
  assertType<res, boolean>();
}
```

## 📃 Motivating Example

Imagine you define an API where the user provides a generator and can interact with your application by the mean of "effects". Something akin to this:

```typescript
run(function* ({call}) {
  const value = yield call(fn);
});
```

Under the current model, TypeScript is forced to type `value` as `any`, because in theory the generator runner may return whatever value it wants to the `yield` expression. In this case however, the result is controlled: we, as developer, know that the runner will return a specific value. All we need is a way to transmit this information to TypeScript.

## 💻 Use Cases

The `redux-saga` library (21.5K ⭐, 6 years old) heavily use generators (there are various reasons why async/await isn't enough, you can see them [here](https://github.com/redux-saga/redux-saga/issues/987#issuecomment-301039792)). As a result, its collection of [effects](https://redux-saga.js.org/docs/api) are all typed as `any`, leaving their users mostly unprotected.

A builtin such as `YieldOverride` would allow them to type their effects in such a way that TypeScript could properly infer the expected return type:

```ts
declare function call<T, Args>(fn: (...args: Args[]) => T, ...args: Args[]): YieldOverride<T>;

function* saga() {
  const val = yield call(() => true);
  assertType<res, boolean>();
}
```

There is a light soundness risk in that the `call` function may also be used in non-saga contexts (in which case its yield override may not be respected). This however seems unlikely, and even in this case the outcome would be almost the same as if the yield operation was to just return `any` (ie the current behaviour).

cc @Andarist

**Comment by nmain** [comment]
Why is this better than using `: Generator<..., ..., T>`?  That can be enforced at call sites too:

[Typescript Playground](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mAVAAmHOAKAlALmQcQFMxCAnAQyjlIB5ywBPAGmQDc4YATFgb2QFs8AZyikYYAObIAvgD5kPAFDJkEBCLbkANsgC8yBjEJbOyACwAmANzLNWgHT8ryAPQvkUBgAdCp8shExSRZyIWRCAA8faF9FaUVFTkIILXJSQlRwaHgwZFJwADF0DDQ4PGw9eSISCipaemY2Dm4FAWFRcSk5XCauG3ywIsxSrGc3ZC9QoUIwz2iAC2SAa2RQ8KjkqFjE5NT0zMhYBDzwACE0jAAjNPKsSoJiMkpqOkYWdi5eZDA8MBB+S5kGSyHofTj9M4XEZjdzAcgwLSzbzJRYQFZrSLRLacRRAA)

Your motivating example could be rewritten to type `run` as something like `run: (f: (props) => Generator<...>) => void` and the type would be inferred inside the anonymous `function*` as well.

**Comment by Andarist** [comment]
@nmain in your demo it's not possible to receive a type dependent on the yielded type. This can be done when using yield delegates (`yield *`) but not with `yield` alone which is used to create "effects" system (like redux-saga, koa etc)

**Comment by nmain** [comment]
I see, yes, that is useful.  Unfortunately your proposal needs a call thunk to do that.  For instance, if my generator runner communicates back a string when yielded a number, and a number when yielded a string, there's no clean way to express that:

```ts
function* foo() {
  // not so pretty
  var a = yield (5 as number & YieldOverride<string>);
  var b = yield ("foo" as string & YieldOverride<number>);
}
```

**Comment by arcanis** [comment]
Exactly, which defers the responsibility of providing the types to the user (at every callsite!), which is what this proposal attempts to avoids. The YieldOverride primitive would be interpreted by TypeScript as "when it goes through a yield expression, this value must resolve to T, regardless what the generator thinks".

The following would thus work as expected: [TypeScript playground](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXwygGsQAeAFQD4AKMACylVRAgC54BhBpliygSnYBNLC2AB5AG4gYMLKD4BuAFChIsBCnTY88MFAgRSAJRAYANPACCMAOYBnGolTtqAOg+wH7Gw4DaALr88AC8lPCmFvAebl72Pnb2gYLwImJSMnIKkZQqquDQcEhomLj49hgwECCo1BUw7PVYqLYpaRAS0rLyZKjIALYARjK5yvnqRWB4FfD2OP0gAMqVzbZcjMxsnNybpE0to2qFCFOoM3MLAHIDwzDrPFv3u31DI3kwaNRapXgAVPDUFIAcRqMigGBwMFIjAAnpZJDh5JYACwAJngAHp-qR4AB1BgYeAAdxwyA68BhpPgAAdkIS6DIQAB+eC-DHhADeyngemmhPqADUDMgECEKaJyYQSHV5ksVi0nix+CoeVB7PYZBgAKIAR2QBj28tsNH4dUqQogIsslRFyrGPNOMxeFpFoXFYgIxBAMquNxkiogdtV6s1uv1RhetxN1GdwpA1o+IDt3N5Z0J1RaGDobphEuAegMEDNVRqlkFcaD8DVGpg2r1BsjI0B1AztizCdtKgAvsmMRj4ABaIfDkejsfjieTqfTmfDsYgAAe1MhhIwMOpCADfDdHIA+rv6BsWOxyF28kcNMVtGV4B9avYoLYoK5gmF4CDmDBwZDoag4fAESRKs-wEdhAOAPI1w3awQ1rMMDXISwtSXcAMBAYBwjFPxEPgZCN0wdCAngRc0NQYB7HgPw8NQ9DLHIIiWRtBB2EQAwNRUb4dHwatQ3rIwAFUTXgLkeTgDBkBgfBOJvUgBRoSQDHYAVLEXfC0OAHxYLrcNSH4yw5OCDkz2ULtlCAA).

**Comment by Andarist** [comment]
That is right but also - maybe not a super common use case. Thunks are usually pretty common for those kinds of patterns.

However, this is a valid criticism and maybe an alternative approach could be figured out. For instance, we could "dispatch" to  the injected type using conditional types like this:

```ts
type Cond<T> = T extends string ? number : T extends number ? string : never;

function* foo(): YieldOverride<Cond> {
  var a: string = yield 5;
  var b: number = yield "foo";
}
```

**Comment by nmain** [comment]
One use case for this is [Mobx's flow](https://mobx.js.org/actions.html#using-flow-instead-of-async--await-), which uses generator functions in which `yield foo` has a runtime type equal to `await foo`.

**Comment by arcanis** [comment]
@andrewbranch I've implemented this feature ([here](https://github.com/microsoft/TypeScript/compare/master...arcanis:mael/yield-override)), do you mind if I open a PR?

**Comment by andrewbranch** [comment]
I think since we haven’t decided on a direction to go for this issue, it’s unlikely that your PR will get much attention from the team, which might mean it gets closed after a while. This issue is still the best place to discuss the problem and try to promote it from “Awaiting More Feedback” to something more committed. In my view, having a prototype implementation doesn’t help its chances, since it doesn’t fall into the category of “it’s a good idea but we don’t know how or don’t have time to implement it.” But generally, it’s not something you need to ask permission for as long as you’re aware that it may not go anywhere. We tend to advise against implementing uncommitted things just to avoid wasted work, but that’s not relevant here since you’ve already done it 😁

**Comment by arcanis** [comment]
Makes sense! What would you say is the current action item, however? Your team hasn't weighted yet, so it's difficult to know what's the type of feedback you're looking for. Is it a case of getting the idea upvoted? Or is there a concrete problem that should be addressed in the proposal?

**Comment by andrewbranch** [comment]
At one point I jotted down this progression:

* Awaiting More Feedback → In Discussion
    * ↑ problem clarity
    * ↑ popularity (volume of 👍 and positive comments)
    * ↑ applicability (breadth of use cases positively impacted)
    * ↑ utility (inverse of the level of pain caused by the problem)
    * ↓ replaceability (other ways of solving the problem)
* In Discussion → Design meeting
    * ↑ solution clarity (everyone understands the high level proposal)
    * ↑ feasibility (we can probably implement it without huge tradeoffs)

I think in this case, the problem clarity is good, the popularity is lacking, the applicability is not yet proven (does anyone outside redux-saga and koa care about this?), the level of pain the problem causes could be elaborated a bit (so all the yield expressions in redux-saga are typed as `any`, but how often do people need to use their results?), and the replaceability of the proposed solution is unclear to me, but I see other ideas being discussed in this thread so I think the design space seems very open still. Maybe @rbuckton can help give some specific guidance, because any shortcomings of Generator typings is something I guarantee he’s thought about a lot.

**Comment by Andarist** [comment]
> so all the yield expressions in redux-saga are typed as any, but how often do people need to use their results?

Most of the time - most of the yielded expressions return something meaningful to the "caller".

**Comment by EmrysMyrddin** [comment]
Another example of library using generators as co-routine is https://github.com/tj/co. It has 11.5k stars, so it's a pretty popular library and has a stable API since a lot of time.

> so all the yield expressions in redux-saga are typed as any, but how often do people need to use their results?

Both library heavily use yield expressions result. You can think of it basically as a plugable `async/await` system which allows to `await` something else than a `Promise`.

**Comment by treybrisbane** [comment]
As much as I'd love a solution to TypeScript's generator typing problems, I don't feel like this is the direction we should go.

The main problem I see with this proposal is that it makes the _yielded value_ responsible for defining a `yield` expression's type. In practice though, it's _the thing that's "running" a generator_ (e.g. [co](https://github.com/tj/co)) that's ultimately responsible for defining a `yield` expression's type (based on the yielded value), because it's the thing that _provides_ a `yield` expression's result (when calling `next`).

IMO, the best way to tackle this is via the approach mentioned by @rbuckton [here](https://github.com/microsoft/TypeScript/pull/30790#issuecomment-480550228), or via the alternative he mentioned [here](https://github.com/microsoft/TypeScript/pull/32695#issuecomment-523733928). Both of these approaches express the type of a `yield` expression as some (type-level) function of the type of a yielded value, which correctly models the runtime behaviour.

**Comment by arcanis** [comment]
> The main problem I see with this proposal is that it makes the _yielded value_ responsible for defining a `yield` expression's type. In practice though, it's _the thing that's "running" a generator_ (e.g. [co](https://github.com/tj/co)) that's ultimately responsible for defining a `yield` expression's type (based on the yielded value), because it's the thing that _provides_ a `yield` expression's result (when calling `next`).

Yes, it would be sounder if the generators were in charge of defining the returned value, there's no argument over that. However, TypeScript's goals explicitly allow for this kind of "incorrectness":

> **Non-goals**
> - Apply a sound or "provably correct" type system. Instead, strike a balance between correctness and productivity.

In this case, the solution you mention is blocked by this other issue https://github.com/microsoft/TypeScript/issues/1213 opened since 2014. While it perhaps will be addressed someday, there's no indication that it'll be anytime soon, or even that it won't suffer from a drawback preventing it from being used in this particular case.

By contrast, letting yielded expressions annotate the value that will be sent back is, admittedly, somewhat backward, but it can be easily implemented with few changes to the codebase (and thus doesn't significantly increase the maintenance cost for the TS team), and solves *most* of the practical use cases seen in popular tools.

It can also be noted that both approaches aren't incompatible: should high order type parameters become a thing, Type Override being a thing won't prevent generators to improve (at worst they'd just become slightly less relevant).

**Comment by arcanis** [comment]
@andrewbranch I've updated the OP with more information and context, does that help?

Something to note is that this issue becomes more important with TS 4.3, because Saga-based projects now yield a bunch of previously non-reported errors, making the migration quite painful (literally hundreds of places that need to be addressed, and the best way for that is to type the returned generator as any, which makes the fix even worse than the problem ...):

```ts
expression implicitly results in an 'any' type because its containing generator lacks a return-type annotation
```

**Comment by andrewbranch** [comment]
@arcanis can you give a small code example of something using redux-saga that gained a new implicit any error in 4.3?

**Comment by arcanis** [comment]
Sure, here's an example: [playground](https://www.typescriptlang.org/play?ts=4.3.2#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXwGcQJwMAeAFQD4AKRVALnhoIygxCalQE8BKeAF4q8CnyYUA3AChQkWAhTpseeAAdkGGlEy5G8bvyYA3HFmAzZ4aHCRpdqgO7mA5iAwEAysVI4YLNg4mADFkCAhPHABbdwALLFQXT0CQcUIMGASXAG0AXRk5G0V7FXw3DAAJLFY-Hho1OGMAQS5eABp4VBBHFoN29UaAIVaeDq7HYb6jeAAlcD9gMlZMxI7lrLyqSwweNQRQ8MiYjHjE5PYEQVn5mEX11fgAb2l4AZAiVAwmZ9fXgCMcLBgEwALJQNRLDJZNZQxJbF6-ZzAcoEJj3HL5BEAX2kWPh0iUDlQACp1MgCLEAHLdc4cCg4Ko1GB1AQ-X5gPCsTrdADqrncBCE8B4WGIwEIPkwNCRKO8JEwfj4Ml+8A5qC5DRAxj5yIFQpFYol8q0CJVrwCFxCYQi0TiWVpqSEVFNZpVrAuADpEDgcB7NR8MB6ZQKXUrpC61VzxoNAbcHfrRRBxURjTQXSqLUF4AcbcdTkkUgJhOlPQCgaHLCrIxg3sYY0D41cDUmjaQ067ze6szmjnazoWnSWOF6fX64AGPWXbhWI5ya-EmTwheVGRhau2O5rtfyPG1079xjqUXuO7X63GUieO9HY8AHTOVVhEMx968F2vmSPfSREid4CIAAZQyeV9hUTcUNBNU8VUeHY9iYAADAhbROLIEJxU8wxVHEcSAA). Seems like it started throwing in 4.2 rather than 4.3, but the point still stands.

This is just one little piece, of course - upgrading required us to add hundreds of `ts-ignore` statements because of this one pattern. The warning itself isn't the problem though, we'd just prefer if we were able to properly type this segment of the codebase in the first place.

**Comment by neurosnap** [comment]
@arcanis [SagaIterator](https://github.com/redux-saga/redux-saga/blob/master/packages/types/types/ts3.6/index.d.ts#L9) is slightly better than typing with `any`.

**Comment by arcanis** [comment]
Ping @andrewbranch? Did the example help?

**Comment by cowboyd** [comment]
Would love to see this gain some traction since all of the use-cases mentioned in the description are _really_ powerful. This is why concepts like structured concurrency are going mainstream in languages like Java and Swift.

In the case of [Effection][] (mentioned in the description) it's one of our biggest points of confusion that trips ups new users, and now, with the dominant mind-share of TypeScript within the JS ecosystem, I'd argue one of the biggest barriers to adoption. In fact, the reason I came upon this issue in the first place, was because I'm typing up a guide on using Effection with TypeScript.

> The main problem I see with this proposal is that it makes the yielded value responsible for defining a yield expression's type. In practice though, it's the thing that's "running" a generator (e.g. co) that's ultimately responsible for defining a yield expression's type (based on the yielded value), because it's the thing that provides a yield expression's result (when calling next).

This is the point exactly. Despite being responsible for providing the value, the exported types that come with "the thing running the generator" currently may have _absolutely no say_ in what the type of that value will be.

However, if you look at all the use-cases above, they pertain to frameworks whose entire _raison d'être_ is to place runtime constraints on how generators will compose. For example, in Effection, the code driving the generator will never ever ever call `next()` with _anything_ but the return value of the invoked generator, the type of which is strictly known. The same is true of both `ember-concurrency` and `redux-saga`, and if it weren't, it would be a critical bug that would garner the full attention of those project's developers. 

We take it as an article of faith that a browser, or the "thing driving the async function" will yield the value produced by a promise and so we consider an `await` expressions type-safe, but we are still putting our trust in that the runtime. Generators by contrast, must always have this "thing running them" that is ever-present, but completely invisible to the type system, and so I'd argue that any type-system that _doesn't_ account for it, or doesn't allow it to become visible is unsound in practice if not in theory.

And in truth, we've seen this happen on our own projects where a manual type annotation becomes "stale" and program that successfully type checks is actually incorrect. On the other hand, allowing the runtime to give the compiler hints about type-inference would actually produce more correct code, more often.

In short, JavaScript has an extension mechanism to the manner in which expressions produce values, but TypeScript does not have a concomitant extension mechanism for expressing the types of those values.

[Effection]: https://frontside.com/effection

**Comment by cowboyd** [comment]
@arcanis I'm assuming `YieldType` would be assignable from any type? I ask because I'm trying to imagine how to write the operation type in Effection so that it still places the right constraints on the actual generators (and not just the yielded values). In other words:

1. If the type parameter is `T`, then the return type of the generator must be `T`
2. the generator may only yield operations and nothing else.

```ts
export type Operation<T> = YieldType<T> & Iterator<Operation<any>, T>;

function* yes(): Operation<boolean> {
  return yield (function*() { return true })();
}

function* no(): Operation<boolean> {
  // @ts-expect-error
  yield 5;
  // @ts-expect-error
  return "false";
}
```

**Comment by andrewbranch** [comment]
@rbuckton would be the best person to give some thoughts on this, though I think he may still be on vacation this week.

**Comment by arcanis** [comment]
@andrewbranch if I open it as a PR, would it be possible to pack it to make it usable in a playground? This way people would be able to play with the proposal.

**Comment by Andarist** [comment]
@arcanis [this PR](https://github.com/microsoft/TypeScript/pull/40468) is packed like this so i assume that it would be possible

**Comment by andrewbranch** [comment]
Yep, feel free to ping me on the PR asking for that 👍

**Comment by arcanis** [comment]
Here's a playground of the feature: [Before](https://www.staging-typescript.org/play?ts=4.5.4#code/PTAEBUAsEsGdQCYFMDGAbAhgJwwF2gPYB2oA7gQK5oKgBGSoADlgQG7TI20CeoABliQIKADwC0sDAHMMfAFDJ02BgDMKRFPmKhYSNKlwAecAD4AFCqIAuUGdi48SGwGUHuJAEpQAXhMQPNgCa0HoI4NyMSMYmANxyciAQMPBwZJB4oLiQDBS6WJhENCgEyGSU1KBoBAQA1pXQNUhy0ETuWCoYKAyujqAA3nKgoKwYaBROoEQUALb0WHEAvvFqGlpEAFSgUkhESDi4BFhmXgNDxUT2OgCOFMoAaqPjPqDcIRW6+pp2bgy+Oj8AOhGYwY602ACYPHEhhhYHlcABRG6jQxTWZ7cweOzIwQPEEAGkyWHGUPiZ2IlzgAHFBI4sFAMEQAIwABmer1COj0Bm+vT+9kcQMeDD8rNJMLhe0RyLQhlo1X0jMxZmptLaDOZLMJuGJnjig1A50u2REeKe3hebxoHx5AvcPj8dqQQpBAIOriwLSkZiZADYPOLQLD4UjbrL7J6iFJlSazUhtbrSUsEmBnARpgxsmhIlh4AdhqMOL1cBEkLA5CXIqAAIKSrDSsPGQkIkSRTRCPwWgDa4GbrYMQgAuqAkCJ3IV4F2W233AhCeBhwB+IlPGwdNC6OJG3BBusN0bPQwAVXxmIdoEMd1PZmBNivI-77YQNlrIZlx8JdxMXj+fQWcSAA) | [After](https://www.staging-typescript.org/play?ts=4.6.0-pr-47548-5#code/PTAEBUAsEsGdQCYFMDGAbAhgJwwF2gPYB2oA7gQK5oKgBGSoADlgQG7TI20CeoABliQIKADwC0sDAHMMfAFDJ02BgDMKRFPmKhYSNKlwAecAD4AFCqIAuUGdi48SGwGUHuJAEpQAXhMQPNgCa0HoI4NyMSMYmANxyciAQMPBwZJB4oLiQDBS6WJhENCgEyGSU1KBoBAQA1pXQNUhy0ETuWCoYKAyujqAA3nKgoKwYaBROoEQUALb0WHEAvvFqGlpEAFSgUkhESDi4BFhmXgNDxUT2OgCOFMoAaqPjPqDcIRW6+pp2bgy+Oj8AOhGYwY602ACYPHEhhhYHlcABRG6jQxTWZ7cweOzIwQPEEAGkyWHGUPiZ2IlzgAHFBI4sFAMEQAIwABmer1COj0Bm+vT+9kcQMeDD8rNJMLhe0RyLQhlo1X0jMxZmptLaDOZLMJuGJnjig1A50u2REeKe3hebxoHx5AvcPj8dqQQpBAIOriwLSkZiZADYPOLQLD4UjbrL7J6iFJlSazUhtbrSUsEmBnARpgxsmhIlh4AdhqMOL1cBEkLA5CXIqAAIKSrDSsPGQkIkSRTRCPwWgDa4GbrYMQgAuqAkCJ3IV4F2W233AhCeBhwB+IlPGwdNC6OJG3BBusN0bPQwAVXxmIdoEMd1PZmBNivI-77YQNlrIZlx8JdxMXj+fQWcSAA) ✨

**Comment by cowboyd** [comment]
Fantastic @arcanis! I [gave it a try][1] on a paired down version of Effection types, and it works marvelously. There was one case I did have a question about. In Effection, `Promise` is a valid operation type, and so you can yield to it, but I couldn't find a way to say "in the context of this generator, if you yield a `Promise<T>`, treat it as though it were `Promise<T> & YieldType<T>`. Obviously don't want to monkey patch the global promise type, but do you think this is the kind of thing that could be inferred in certain circumstances?

[1]: https://www.staging-typescript.org/play?ts=4.6.0-pr-47548-5#code/C4TwDgpgBA8pBOBDYBLA9gOwDwBUB8UAvFABQCSwESwa8WcVy62iGIeANFDlwK4YBrDGgDuGAgB8oABXhoAtigDOEXHgCUUAGRQAmiggAbACY5wq-AG4AUNYBm-AMapMAKiiYIJdQC5YCJkwsAEYCAG9rKCh4CGBeeAwoYJsAX1sHDGdmd2ARNG8-BmpmLAAmcMjo2PjE0tT0pxcMHIALGK9ff0YmrAxeeQAjKnCoysNYqqVeQ2AiUhADEw8MDs0AannF4yhc-PV1GyiYuITJ6eB6+0bsnYgMAq7ioL7B4agIqPHZlQA3O7mFkZtrIFMoIAA6GJKNCGP4kADsB0qxxqUF+-w2JEBS2AbQgq1SQA

**Comment by Andarist** [comment]
It's an inherent trait of this design - it's the yielded value/type that decided what gets "injected" back to the generator. So, as you have said, it's somewhat impossible to overload existing types with a custom/unwrapped/whatever yielded type. So to make it work you kinda have to convert the promise using a wrapper function, like [here](https://www.staging-typescript.org/play?ts=4.6.0-pr-47548-5#code/C4TwDgpgBA8pBOBDYBLA9gOwDwBUB8UAvFABQCSwESwa8WcVy62iGIeANFDlwK4YBrDGgDuGAgB8oABXhoAtigDOEXHgCUUAGRQAmiggAbACY5wq-AG4AUNYBm-AMapMAKiiYIJdQC5YCJkwsAEYCAG9rKCh4CGBeeAwoYJsAX1sHDGdmd2ARNG8-BmpmLAAmcMjo2PjE0tT0pxcMHIALGK9ff0YmrAxeeQAjKnCoysNYqqVeQ2AiUhADEw8MDs0AannF4yhc-PV1GyiYuITJ6eB6+0bmKABzWIA5foKu4qC+weGoCKPq09kFMoIAA6GJKNCGABuXgA7OprGkrpkmjkIBgXkVAtgPkN4BUouNZipoYliAsjNt7sAnvJvJVjjUoMS0VANiRyUtgG0IKtUkA) (I'm pretty sure that you know that, I'm just posting this as a reference for others).

Note that with your `Operation` type I've expected an error [here](https://www.staging-typescript.org/play?ts=4.6.0-pr-47548-5#code/C4TwDgpgBA8pBOBDYBLA9gOwDwBUB8UAvFABQCSwESwa8WcVy62iGIeANFDlwK4YBrDGgDuGAgB8oABXhoAtigDOEXHgCUUAGRQAmiggAbACY5wq-AG4AUNYBm-AMapMAKiiUMJdQC5YCJkwsDF55ACMqAgBvaygoAHp4qAABYCUAWggAD0hnTPg5eFioeAhgXngMKAByAEYABmrrAF8gA) but it doesn't raise one 🤔

**Comment by cowboyd** [comment]
> It's an inherent trait of this design - it's the yielded value/type that decided what gets "injected" back to the generator

@Andarist Right, but in this case because the `TYield` type of my operation is __itself__ `Operation`, so in order for the generator function to type check, you cannot yield `Promise<T>` because that is not an operation, however `Promise<T> & YieldType<T>` is an operation. So I was wondering if that could then be used to invert the yield type of promises. 

It's a bit circular, but what I'm wondering is why it can't figure out from the type of the generator that the yielded value is actually `Promise<T> & YieldType<T>`(which can therefore be used to infer the left hand side of the expression.

**Comment by Jack-Works** [comment]
I support this as a worth-solving problem. Yield-generator itself in JavaScript can be used to emulate powerful control flow structures (as many examples described above). Lacking type support will limit the user experience in this case.

I have a use case https://github.com/Jack-Works-forks/rfcs/blob/cancel-async-effects/text/0000-cancel-async-effects.md#basic-example I proposed an auto-cancellation effect API but it is lacking support from the type system.

```js
useEffect(function* (_signal) {
    setLoading(true)
    // React automatically stops if the deps array outdated.
    setText(yield getData(props.id))
}, [props.id])
```

In the case above, `yield T` will be always `T` by the semantic of `useEffect` I proposed. But there is no way to modal it in the type system.

I believe `YieldType<T>` is not enough, it would be better if the type system can behave this way:

```ts
type UnboxPromise<T> = T extends PromiseLike<infer Q> ? Q : never
declare function co<T, TR>(g: () => Generator<T, TR, UnboxPromise<T>>): Promise<TR>

// contextual typing for the generator function
co(function* () {
    // T = Promise<number>, UnboxPromise<T> results in number
    const a = yield Promise.resolve(1)
    //    a: number

    // T = Promise<string>, UnboxPromise<T> results in string
    const b = yield Promise.resolve("")
    //    b: string

    return Promise.resolve(true)
}) //return: Promise<boolean>
```

This requires the type system to specially handle generator function to provide an ad-hoc multiple type evaluation.

**Comment by arcanis** [comment]
> @rbuckton would be the best person to give some thoughts on this, though I think he may still be on vacation this week.

Ping @rbuckton, do you have any thoughts on this?

**Comment by nightlyherb** [comment]
In response to this concern:
> Given the discussion there, I personally feel that this is not the right approach—this is essentially just a type assertion, but piggy-backing on non-type-assertion syntax in such a way that nobody will recognize that it’s just as unsafe. The next step is not finding someone to champion this PR; we are still on the very first step: exploring possible solutions that could be driven to consensus.
https://github.com/microsoft/TypeScript/pull/47548#issuecomment-1167579213

I am not sure but I think I found one way of implementing a type-safe method for annotating and calling generators:

```ts
// This does not contradict the type system, hopefully?

interface MonadicGenerator<YieldCount extends number, YieldList extends unknown[], Return extends any, NextArgsList extends unknown[][]> {
    next<Index extends YieldCount>(...args: NextArgsList[Index]): IteratorReturnResult<Return>;
    next<Index extends number = number>(...args: NextArgsList[Index]): IteratorYieldResult<YieldList[Index]>;
}

const genFunc = function* () {
    const a = yield new Just(5);
    const b = yield new Just("6");
    return new Just(a + parseInt(b, 10));
} as () => MonadicGenerator<2, [Maybe<number>, Maybe<string>], Maybe<number>, [[undefined], [number], [string]]>

function naiveDoNotation<T0, T1, R> (
    genFunc: () => MonadicGenerator<2, [Maybe<T0>, Maybe<T1>], Maybe<R>, [[undefined], [T0], [T1]]>
) {
    const gen = genFunc();
    const { value: ma } = gen.next<0>(undefined);
    return ma.bind((a) => {
        const { value: mb } = gen.next<1>(a);
        return mb.bind((b) => {
            const { value: result } = gen.next<2>(b);
            return result;
        })
    })
}
```
Link to full playground: https://github.com/microsoft/TypeScript/issues/36967#issuecomment-1241492260

I was not aware of this issue here when I was writing this code so this demo doesn't use YieldType
but maybe this idea can be incorporated to this proposal somehow?

**Comment by wnadurski** [comment]
Following this blogpost: https://www.matechs.com/blog/abusing-typescript-generators
I have implemented a generic do notation that makes use of fp-ts implementation of HKT and type classes.
This is how it can be used:

```typescript
it('works for option', () => {
  const result = Do(O.Monad)(function* (_) {
    const x = yield* _(O.some(2))
    const y = yield* _(O.some('asd'))

    return y + x.toString()
  })

  expect(result).toEqual(O.some('asd2'))
})

it('works for IO', () => {
  const someSideEffect = jest.fn()

  const result = Do(IO.Monad)(function* (_) {
    const x = yield* _(IO.of(2))
    yield* _(() => someSideEffect())

    return 'asd' + x.toString()
  })

  expect(result()).toEqual('asd2')
  expect(someSideEffect).toHaveBeenCalled()
})

it('works for Task', async () => {
  const someSideEffect = jest.fn().mockReturnValue(Promise.resolve(undefined))

  const result = Do(T.Monad)(function* (_) {
    const x = yield* _(T.of(123))

    yield* _(someSideEffect)

    const y = yield* _(T.of(10))

    return x + y
  })

  expect(await result()).toEqual(133)
  expect(someSideEffect).toHaveBeenCalled()
})
```
Here is full code: https://gist.github.com/wnadurski/51a3e8cf06c8fe0cc3dd50b87fd90365

**Comment by tadhgmister** [comment]
> Note that with your `Operation` type I've expected an error [here](https://www.staging-typescript.org/play?ts=4.6.0-pr-47548-5#code/C4TwDgpgBA8pBOBDYBLA9gOwDwBUB8UAvFABQCSwESwa8WcVy62iGIeANFDlwK4YBrDGgDuGAgB8oABXhoAtigDOEXHgCUUAGRQAmiggAbACY5wq-AG4AUNYBm-AMapMAKiiUMJdQC5YCJkwsDF55ACMqAgBvaygoAHp4qAABYCUAWggAD0hnTPg5eFioeAhgXngMKAByAEYABmrrAF8gA) but it doesn't raise one thinking

@Andarist This is just a limitation of typescript where if the return type of a generator has a conditional non iterator type it can't check the yield sites. ([playground](https://www.staging-typescript.org/play?ts=4.9.5#code/FAMwrgdgxgLglgewgAgFQwKYGcYAoCUAXMrgJKYBOAhjAhQDw4VwQDmAfMgD7IAKFCALZwsGegHZ2+ZAG9gyZAE84GADYATZAEYADMAC+QA)) it has nothing to do with this proposal and is present in the latest version of typescript.

>In the case above, yield T will be always T by the semantic of useEffect I proposed. But there is no way to modal it in the type system.
>
>I believe YieldType<T> is not enough, it would be better if the type system can behave this way:

@Jack-Works you are basically describing #36967, in fact with parameter inference you could probably specify the expected `yield` parameter in your library function and have it automatically inferred for the generator function passed directly in the same way that inference works [here](https://www.staging-typescript.org/play?ts=4.6.0-pr-47548-5#code/JYOwLgpgTgZghgYwgAgIIG8BQzlwFzIDOYUoA5pgL6YwCuICYwA9iMmHANYSECyEYABbMAJgHVgQgCqDghVFDIAKGASVC5BVAEoAvAD4Abs2AjtWHDAB0COABs7S9PgDkgiA+YvK2qpkwc3HwCwuKSgjJyCsp0DEysSubYyAishMx2EFZ2zMoahFZwvj5AA)

> I couldn't find a way to say "in the context of this generator, if you yield a `Promise<T>`, treat it as though it were `Promise<T> & YieldType<T>`.

@arcanis I don't see how this actually fixes anything unless you are specifically using a library that maintains complete control over the type at every single yield site, if my library just uses `Promise` or even strings or similar objects at the yield site then you still end up needing some way to inject the `YieldType` into the type within the generator.

Like pretend I had a library that the yielded values were just strings to denote labels in a database that is lazily loaded as needed and the list of valid labels is statically known, I don't think this proposal actually benefits this.
```typescript
interface A { a: number }
interface B { b: string }
function* current(){ 
  // doesn't actually ensure the "A"->A, "B"->B logic
  const x: A = yield "A" as const
  const y: B = yield "B" as const
  assertType<A>(x); assertType<B>(y);
}
function* using43632(){
  // unclear whether this is actually any better...
  const x = yield ("A" as "A" & YieldType<A>);
  const y = yield ("B" as "B" & YieldType<B>);
  assertType<A>(x); assertType<B>(y);
}
```
This is obviously a toy example but this is exactly analogous to the case of yielding `Promise`s which is very much a desired usecase. 

I feel like an approach that decouples the library yield contract from the type of yielded value is a better approach such as #36967 ([playground](https://www.staging-typescript.org/play?ts=4.6.0-pr-47548-5#code/FASwdgLgpgTgZgQwMZQAQEFUG9itQgLlTAFcBbAI1mAF9RJZEVUAhbXVCogZwhnADmtYHBJgkEEAHswAKlRISMGFEgAKAJRZUHAPS7UAEylRuYAOQR8EkggA2dgJ6pV3JWggALNACJ0PgFoAPnQAGlQfFkCgtjspARAkDiQZXlQADyJMAF5URxAoO0MI-3xuBVSIZMq8ojZc-MLiyJ8yirBeDgRublgIABVHAAcoAB50ILV0jQBuMt6YAeGxlknHWeFRcUkZeRJuQQAWAGYANmOAJk0cPH1UMSQ7KAQYVAB3by9YVC8Qcr-rBBbA5nAgwM4qBBoDAAHRw6odKzpVANApFVBqPytbolVoAMlQAE00YZBiNxkENngUoi8ii8iSMS02syCcSmmSVpSZl0en1ORSprN5vzlqNVmp1jy6HcIMtUAAZEAUGAvRwAESgcHAUEM7KKAGEZHxkFZclg1HYEFQ7AQsRoCOg5pbrYU7VEHSwZQYthJpHJ7gcwAIzgBOU4AdklJIISpVas12rAuv1hiNkFVEi0wDu1JqyNRTVxOYMedpzkL6JaJbwIsWAomQrm3QWS3JEqlNbowGAhigjxeaF9OzAdbbY36k0y-QdADcpCBDEA) showing equivalent code in both cases)

Note that in your PR the `TSend` generic is currently evaluating to `unknown` which makes it really easy to write an unsound runner code. in the playground above `current` has a TSend value of `A&B` since that is the only type that'd satisfy all yield positions so a runner would need to either satisfy all of them or use type assertions, this is basically what I get at in https://github.com/microsoft/TypeScript/issues/36855#issuecomment-589081258

**Comment by uhyo** [comment]
Hello, maybe I'm a skosh late to the party, but I hope you find below information useful.

## Another Motivation for an Enhanced `yield` Typing

Recently, [Jotai](https://jotai.org/) is getting popular among state management libraries for React. [Compare Jotai and redux-saga by npm downloads](https://npmtrends.com/jotai-vs-redux-saga); Jotai overtook redux-saga this year.

Jotai promotes an ″[async sometimes](https://jotai.org/docs/guides/async#async-sometimes)” pattern, which essentially means to deal with values that are sometimes sync and sometime async (in other words, of type `T | Promise<T>`). This pattern has been proven practically important to build a good UI with React.

To deal with such values, the [soon](https://jotai.org/docs/third-party/derive#soon) function from `jotai-derive` is useful. This function  applies given callback to given sometimes-async value, keeping syncness if possible. A simplified version of the `soon` function looks like:

```ts
function soon<T, U>(value: T | Promise<T>, callback: (value: T) => U | Promise<U>): U | Promise<U> {
  if (value instanceof Promise) {
    return value.then(callback);
  }
  return callback(value);
}
```

This can be thought of as a generalized version of Promise#then that keeps syncness of given value if possible.

Then, what is the “async sometimes” version of async functions? You know, you can build such a thing using generator functions.

```ts
function runSoon<R>(callback: () => Generator<any, R, any>): R | Promise<R> {
  const generator = callback();
  return rec(null);
  
  function rec(current: any): R | Promise<R> {
    const { done, value } = generator.next(current);
    if (done) {
      return value;
    }
    return soon(value, rec);
  }
}

const result1 = runSoon(function*() {
  const value1 = yield 1;  // value1 === 1
  const value2 = yield 41; // value2 === 41
  return value1 + value2;
});
console.log(result1); // result1 === 42

const result2 = runSoon(function*() {
  const value1 = yield Promise.resolve(1); // value1 === 1
  const value2 = yield Promise.resolve(41); // value2 === 41
  return value1 + value2;
})
console.log(result2); // result2 === Promise { value: 42 }
result2.then((v) => console.log('result2 is', v));
```

It would be great if TypeScript can typecheck and infer types for code like above!

## Another Proposed Solution

The idea is that the `yield` expression can be considered as a pseudo function call for the purpose of type checking.

Let's consider a new interface for generator functions:

```ts
interface EnhancedGeneratorFunction<TArgs, TYield, TReturn>
```

Then, `runSoon` could be typed as follows:

```ts
function runSoon<R>(
  callback: EnhancedGeneratorFunction<[], <T>(value: T) => Awaited<T>, R | Promise<R>>
): R | Promise<R>
```

Then, the generator function passed to `runSoon` is contextually typed, and `yield` expressions are typechecked like a call to a function of type `TYield`.

```ts
const result1 = runSoon(function*() {
  const value1 = yield 1;  // ← this can be checked as if passing 1 to a function of type TYield
  const value2 = yield 41;
  return value1 + value2;
});
```

Pros of my proposal are:

- enables contextual typing.
- flexible enough to support various use cases.

Cons are:

- Not clear how to declare the `TYield` type inside a generator function (when no contextual typing). Maybe requires a new syntax.

**Comment by tadhgmister** [comment]
@uhyo your proposed solution is effectively #36967.
